File: arrays/contains_duplicate_ii.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Boolean}
def contains_nearby_duplicate(nums, k)
  map = {}
  nums.each_with_index do |num, idx|
    return true if map.key?(num) && (map[num] - idx).abs <= k
    map[num] = idx
  end

  false
end

# 219. Contains Duplicate II
# https://leetcode.com/problems/contains-duplicate-ii

File: hash_table/contains_duplicate_ii.py
from typing import List


class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        num_map = {}

        for i, num in enumerate(nums):
            if num in num_map and abs(i - num_map[num]) <= k: return True
            num_map[num] = i

        return False


# 219. Contains Duplicate II
# https://leetcode.com/problems/contains-duplicate-ii


sol = Solution()
assert sol.containsNearbyDuplicate([1, 0, 1, 1], 1), True
assert sol.containsNearbyDuplicate([1, 2, 3, 1], 3), True
assert sol.containsNearbyDuplicate([1, 2, 3, 1, 2, 3], 2), False
